### PR TITLE
Enhance/masterfile

### PR DESCRIPTION
**Pull request by 1984Doc**
### **User description**
Start of changes to master file.  Once merged will clean up edit etc.  

## Checklist

- [x] My commits are signed off for the [DCO](https://developercertificate.org/) (`git commit -s`)
- [x] My commits follow [Conventional Commits](https://www.conventionalcommits.org/) format, or I've written clear commit messages and will use the format next time
- [x] I have not included any patient data (PHI) in this PR
- [x] I have added tests for new functionality, or this change doesn't need new tests
- [x] I have read the [contributing guide](CONTRIBUTING.md)

___

### **Description**
- This PR modernizes the demographic edit page to align with the CARLOS design language.
- It includes a complete overhaul of the layout and styling, introducing a full-width patient banner, sticky header, and streamlined sidebar.
- The bottom toolbar has been reorganized for better accessibility, and a toggle feature for hiding empty fields has been added.
- Additionally, the search functionality has been enhanced with a toggle link for improved user experience.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>demographiceditdemographic.css</strong><dd><code>Modernized CSS for Demographic Edit Page</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/webapp/demographic/demographiceditdemographic.css
<li>Overhauled layout and styling for the demographic edit page.<br> <li> Implemented full-width patient banner and sticky header.<br> <li> Streamlined sidebar and reorganized bottom toolbar.<br> <li> Added Show/Hide toggle for empty fields.<br>

</details>

  </td>
  <td><a href="https://github.com/carlos-emr/carlos/pull/654/files#diff-2f6d790b97b584fa13d693f977168c19e791dd448fb676df8ea1d27bd2823773">+792/-84</a></td>
</tr>

<tr>
  <td>
    <details>
      <summary><strong>demographiceditdemographic.jsp</strong><dd><code>Updated Demographic Edit Page Structure</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/webapp/demographic/demographiceditdemographic.jsp
<li>Integrated new CSS file for styling.<br> <li> Updated patient header with improved layout.<br> <li> Added search toggle functionality.<br> <li> Enhanced toolbar with new buttons and dropdowns.<br>

</details>

  </td>
  <td><a href="https://github.com/carlos-emr/carlos/pull/654/files#diff-343b10ea34d325c50823f447ab1cc82e9ae762408ff5c7aec7fefb1878915e69">+172/-258</a></td>
</tr>

<tr>
  <td>
    <details>
      <summary><strong>zdemographicfulltitlesearch.jsp</strong><dd><code>Refined Search Full Title Page</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/webapp/demographic/zdemographicfulltitlesearch.jsp
<li>Modified back button to use a specific class for styling.<br> <li> Improved button functionality for better user experience.<br>

</details>

  </td>
  <td><a href="https://github.com/carlos-emr/carlos/pull/654/files#diff-8c35a1dafdf9015d55dc22c63aadaa5b54b363a3e972d24fb3998869eb1d88da">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **Penify usage**:
>Comment `/help` on the PR to get a list of all available Penify tools and their descriptions

## Summary by CodeRabbit

* **New Features**
  * Toggle to hide/show empty demographic fields, embedded patient search toggle, compact print/labels dropdown, consolidated toolbar with provider actions and Care Connect, updated patient header and top-status components.

* **Style**
  * Complete visual redesign of Demographic Edit to CARLOS flat theme: header, left navigation, two-column layout, panels, forms/tables, buttons, tooltips, sticky top row, responsive wrappers and print rules.

* **Localization**
  * Added UI labels for demographic editing and expanded eForm translations across multiple locales.

### CODE REVIEW

**Comment by 1984Doc** [comment]
@claude review

**Comment by phc007** [comment]
@claude fix all bot identified issues
check and fix OWASP coverage
check and fix i18n coverage adding key values to each of the language resource files
then advise if it LGTM

**Comment by phc007** [comment]
@claude 
The !important CSS (~350 uses) is being used to override encounterStyles.css and has multiple side effects
and is difficult to maintain
Instead drop the reference to encounterStyles.css and drop the !important where it was used as override to encounterStyles.css

<%@ include file="/includes/global-head.jspf" %> is the approach used in CARLOS to ease maintenance of common libraries to allow for upgrades across the project when you change in one spot.  Currently this includes
    - Viewport meta tag for responsive design
    - global.js (legacy focus/refresh helpers)
    - jQuery 3.7.1
    - Bootstrap 5.3.3 (JS bundle + CSS)
    - jQuery UI 1.14.2 CSS
    - Font Awesome 6.7.2 (icon library)
    - searchBox.css (shared search/form styles)
    - global.css (CARLOS design tokens and common classes)
    - carlos-ajax.js (CARLOS shim for remaining prototype.js functions)
These are the libraries and approaches used as a default and it defeats the purpose to reference them separately directly in a jsp.  Remove references in this work that duplicate or override those in global-head.jspf

Thereafter Verify each finding against the current code and only fix it if needed.

Inline comments:
In `@src/main/resources/oscarResources_fr.properties`:
- Line 4484: Update the French label value for the key
demographic.demographiceditdemographic.searchPatient to use correct phrasing by
replacing "Rechercher patient" with "Rechercher un patient" in
src/main/resources/oscarResources_fr.properties so the UI shows the
grammatically correct French string.

In `@src/main/webapp/demographic/demographiceditdemographic.css`:
- Line 31: The universal rule "body#demographiceditdemographic * { font-family:
... !important; }" is too broad and breaks icon-fonts; replace it with a scoped
rule targeting only text-bearing elements (e.g. body#demographiceditdemographic,
body#demographiceditdemographic p, h1-h6, label, span, a, li, td, th, input,
textarea, button, select) and remove the !important so icon classes (and
injected icon elements) keep their own font-family; update the selector that
currently reads "body#demographiceditdemographic *" accordingly.
- Around line 419-423: The current CSS rule using table#editDemographic tr td
input (and select/textarea) applies width:100% to all input types (including
checkboxes, radios, file, hidden), causing layout issues; update the selector to
target only text-like inputs plus textarea/select (for example replace the
generic input with input[type="text"], input[type="email"],
input[type="password"], input[type="search"], input[type="tel"],
input[type="url"], input[type="number"], input[type="date"] along with textarea
and select) so only those controls receive width:100% and font-size styling
while leaving checkboxes/radios/files unaffected; modify the rule referencing
table#editDemographic tr td input/select/textarea accordingly.
- Around line 402-417: The CSS forcibly converts table layout to flex for
table#editDemographic and its nested table#waitingListTable (selectors
table#editDemographic tr, table#editDemographic tr td table#waitingListTable tr
and their td rules), breaking native table semantics and colspan handling;
revert these rules to use display: table-row and display: table-cell (remove
display:flex and !important overrides), restore width/margin behavior
appropriate for table cells (use width:100% only where needed), and replace flex
alignment properties with table-friendly properties like vertical-align and
text-align so the colspan-based layout in `#editDemographic` and `#waitingListTable`
continues to work.

In `@src/main/webapp/demographic/demographiceditdemographic.jsp`:
- Around line 1237-1244: The rendered user-derived variables head,
records.get(i) and demographic_no must be encoded per output context: update the
anchor href query values (the demographic_no in the URL in the <a> tags) to use
Encode.forUriComponent(...) and update the visible text portions (e.g., "#<%=
head %>", "#<%= records.get(i) %>", "#<%= demographic_no %>") to use
Encode.forHtml(...); also locate the JavaScript onclick usage at the other
occurrence (around line 1598) and wrap the value with
Encode.forJavaScriptAttribute(...). Apply these changes to the <a href="...">
query parameter values, the inner HTML text nodes, and the JS onclick attribute
to ensure context-appropriate encoding using Encode.forUriComponent,
Encode.forHtml, and Encode.forJavaScriptAttribute respectively.

---

Duplicate comments:
In `@src/main/webapp/demographic/demographiceditdemographic.jsp`:
- Line 1000: Replace the raw stored sex code rendering
Encode.forHtml(demographic.getSex()) with the human-readable display label used
elsewhere (for example use demographic.getSexLabel() or
Sex.getLabel(demographic.getSex()) depending on the existing mapping utility),
keeping the Encode.forHtml(...) wrapper so the displayed label is HTML-escaped;
update the patient header span to call that display-mapping method instead of
demographic.getSex().
- Around line 249-250: Remove the duplicate viewport meta tag rendered in
demographiceditdemographic.jsp (it appears after the <%@ include
file="/includes/global-head.jspf" %> include and again near the title block);
since global-head.jspf already outputs a viewport, delete the additional <meta
name="viewport" ...> occurrence (also present around the later block mentioned)
so only the include-provided viewport remains to avoid conflicting mobile
scaling behavior.

**Comment by phc007** [comment]
@claude Verify each finding against the current code and only fix it if needed.

Inline comments:
In `@src/main/webapp/demographic/demographiceditdemographic.css`:
- Around line 48-49: The CSS uses color-mix() in several rules (e.g., selector
body#demographiceditdemographic a:hover) which may lack safe fallbacks for older
browsers; for each occurrence of color-mix (lines shown in review) add an
explicit solid fallback color declaration immediately before the color-mix
declaration so browsers that don’t support color-mix pick the fallback, e.g.,
add a color: <hex-fallback>; line before color: color-mix(...); to every rule
that currently uses color-mix (search for color-mix in the file and update each
matching selector such as body#demographiceditdemographic a:hover and the other
instances listed).

In `@src/main/webapp/demographic/demographiceditdemographic.jsp`:
- Line 4877: Replace the URL encoding call inside the popupPage invocation so it
uses the project's OWASP encoder instead of java.net.URLEncoder: locate the
popupPage(...) call that builds the query parameter appointment_no (the line
containing "<%=printLbl%><%=demographic.getDemographicNo()%>&appointment_no="
and URLEncoder.encode(..., StandardCharsets.UTF_8)) and change the encoding
expression to use Encode.forUriComponent(appointment) (or
Encode.forUriComponent(appointment != null ? appointment : "")) to match the
pattern used elsewhere in this JSP; ensure the new expression is wrapped in the
same JSP expression tags so the popupPage string is built the same way.
- Around line 4853-4860: The BC_CARECONNECT_URL value is injected raw into the
JavaScript onclick string for callCareConnect, which can break the JS if the URL
contains quotes or backslashes; update the onclick call to pass an encoded URL
the same way BC_CARECONNECT_REGION is encoded (use
Encode.forJavaScriptAttribute(...) around
OscarProperties.getInstance()['BC_CARECONNECT_URL'] or the ${url} variable) so
callCareConnect receives a safely escaped JavaScript string; ensure you only
change the URL argument and keep the other encoded demographic.* arguments
intact.

---

Nitpick comments:
In `@src/main/webapp/demographic/demographiceditdemographic.css`:
- Around line 723-727: There are two conflicting .darkPurple CSS rules;
consolidate the properties into the original .darkPurple declaration (the one
using background-color: var(--carlos-bg-light, `#f5f5f5`)) 

Verify each finding against the current code and only fix it if needed.

Inline comments:
In `@src/main/resources/oscarResources_es.properties`:
- Line 5029: The label demographic.demographiceditdemographic.btnCareConnect
currently uses the English string "CareConnect";  append a Spanish descriptor as "CareConnect"
is a service
name (e.g., "CareConnect (servicio)"), ensuring the
value matches the locale style and surrounding labels while leaving the product
name unchanged if required.
- Line 5027: Update the UI label value for the key
demographic.demographiceditdemographic.searchPatient to use natural Spanish
phrasing by changing its value from "Buscar Paciente" to "Buscar un paciente" so
the resource entry reads:
demographic.demographiceditdemographic.searchPatient=Buscar un paciente.

---

Duplicate comments:
In `@src/main/webapp/demographic/demographiceditdemographic.jsp`:
- Around line 1243-1250: The toggle helpers (showHideDetail(), showEdit(),
showHideItem()) currently assume elements like swipeButton exist and restore
hidden elements using display: table, which breaks when workflow_enhance is off
and when new nodes (swipeButtonBottom, <div id="viewDemographics2">) are
present; update these helpers to null-safe access (check element !== null before
touching .style) and include logic to handle swipeButtonBottom and
viewDemographics2 alongside swipeButton, and when restoring visibility use
appropriate display values (block or inline as applicable for divs and inline
elements) instead of always using "table" so the read view layout returns
correctly after toggling.
- Around line 4875-4877: Replace the use of URLEncoder.encode(...) in the
popupPage call with OWASP Encoder's Encode.forUriComponent(...) to consistently
apply context-specific URI-component encoding; specifically, in
demographiceditdemographic.jsp change the onclick that currently uses
URLEncoder.encode(appointment != null ? appointment : "",
StandardCharsets.UTF_8) to use <%=
org.owasp.encoder.Encode.forUriComponent(appointment != null ? appointment : "")
%> (or ensure org.owasp.encoder.Encode is imported/accessible) so the
appointment_no query param is encoded with Encode.forUriComponent in the
popupPage(...) invocation.
- Around line 4852-4859: The page injects the page-scoped variable url directly
into the onclick JS string for callCareConnect, which can break if the URL
contains quotes or backslashes; fix by encoding the URL with the OWASP encoder
(use Encode.forJavaScriptAttribute(url) when rendering the url into the onclick)
or alternatively place the raw url into a data-* attribute and read it in the
onclick handler; update the callCareConnect invocation around the existing
callCareConnect(...) usage so the first argument is the encoded value instead of
${url}.

Is this LGTM

**Comment by yingbull** [comment]
@claude the audit information option that was removed, can you put it back in

**Comment by yingbull** [comment]
Yes, really need screenshots to help on these very visual ones @1984Doc.   Thank you @phc007 for all your efforts here, your efforts and expertise are welcome.

**Review by yingbull — APPROVED** [review]
